documentation/general/dash-sai-pipeline-packet-flow.md

```diff
@@ -0,0 +1,799 @@
+
+```mermaid
+flowchart TD
+    subgraph ENI
+        EL[ENI Lookup]
+
+        EP0[ENI 0 Pipeline]
+        EP1[ENI 1 Pipeline]
+        EP2[ENI 2 Pipeline]
+    end
+
+    subgraph LoadBalancer
+        VL[Vip Lookup]
+
+        LBP0[LB 0 Pipeline]
+        LBP1[LB 1 Pipeline]
+        LBP2[LB 2 Pipeline]
+    end
+
+    EL --> |ENI 0| EP0
+    EL --> |ENI 1| EP1
+    EL --> |ENI 2| EP2
+
+    VL --> |LB 0| LBP0
+    VL --> |LB 1| LBP1
+    VL --> |LB 2| LBP2
+```
+
+A pipeline can also define its initial matching stages, which will be used for starting processing the packets if no flow is matched.
+
+> NOTE:
+> 
+> DASH-SAI pipeline is a logical concept. It doesn't have to be 1 to 1 mapping to a physical ASIC pipeline. The underlying implementation can be as simple as a metadata field update with a specific value, when the entry getting matched, e.g., `ENI = Inner Source/Destination MAC`.
+
+### 5.4. Packet Decap
+
+If a pipeline is found, before processing the packets, all outer encaps will be decap'ed, and with key information saved in metadata bus, such as encap type, source IP and VNI / GRE Key, exposing the inner most packet going through the pipeline. This simplifies the flow matching logic and also allow us to create the reverse flow properly.
+
+#### Stateless decap vs stateful decap
+
+During the direction lookup stage, all other encaps will be examined, such as VNI lookup. For each VNI or GRE key, we can specify whether it is stateless or stateful.
+
+Although the encap information will still be saved in the metadata bus, however, for stateless decap, we will not create the encap for the reverse flow. Otherwise (for the stateful decap), we will create the encap for the reverse flow by reversing the source and destination information.
+
+```json
+{
+    "DASH_SAI_VNI_TABLE|12345": { "direction": "outbound", "stateless": false },
+    "DASH_SAI_VNI_TABLE|12346": { "direction": "outbound", "stateless": true }
+}
+```
+
+### 5.5. Conntrack Lookup and Update
+
+After entering a specific pipeline, the first stage will be the Conntrack Lookup stage, which does the flow lookup. If any flow is matched, the saved actions will be applied, the metering counters will be updated, and the rest of pipeline will be skipped.
+
+The core of the Conntrack Lookup and Update stage is the flow table.
+
+#### 5.5.1. Flow lookup
+
+First, let's define the flow lookup behavior.
+
+The flow lookup **MUST** use the information of inner most packet. And the matching keys can be configurated via the DASH flow APIs during the DASH initialization. This allows us to support different flow matching behaviors for certain cases. For example, in L3 level routing, we don't need to create 5 tuple flows.
```

Comment:
If I understand correctly, you are not saying that arbitrary fields can be used as the flow keys. Instead, from the packet there is at most the classic 5-tuple fields. And any of these 5 fields can be disabled from the key based on a config given at DASH init time? (To disable the field basically means using a fixed value like 0).

Reply:
Hi Bud, thanks a lot for review and leaving comments! For your questions:

> If I understand correctly, you are not saying that arbitrary fields can be used as the flow keys. Instead, from the packet there is at most the classic 5-tuple fields.

Yes, modeling the arbitrary fields as flow key will be very complicated, and also hard to implement on certain DPU/ASICs. So, the classic 5 tuple besides certain frequently used fields should be a place to start with. More details can be found here in the flow api spec (in review): See `sai_dash_flow_key_t` in <https://github.com/sonic-net/SONiC/blob/d2bb1913dacf682cfa2affc0723101df5fc6ed81/doc/dash/dash-flow/DASH_FLOW_SAI.md>. 

> And any of these 5 fields can be disabled from the key based on a config given at DASH init time? (To disable the field basically means using a fixed value like 0).

Yes, and yes, using 0 is actually a valid way to implement this. There are certain DPUs that actually allow the stages to be dynamically created too, so the underlying implementation is flexible, depending on the architecture.